src/memprof_tracer.ml

```diff
@@ -104,68 +94,69 @@ let start ?(report_exn=default_report_exn) ~sampling_rate trace =
         match Trace.Writer.put_collect trace (Trace.Timestamp.now ()) id with
         | () -> unlock_tracer s
         | exception e -> mark_failed s e) } in
-  curr_active_tracer := Some s;
-  bytes_before_ext_sample := draw_sampler_bytes s;
-  Gc.Memprof.start
+  Atomic.set curr_active_tracer (Some s);
+  Atomic.set bytes_before_ext_sample (draw_sampler_bytes s);
+  let _t = Gc.Memprof.start
     ~sampling_rate
     ~callstack_size:max_int
-    tracker;
+    tracker in
   s
 
 let stop s =
   if not s.stopped then begin
     s.stopped <- true;
     Gc.Memprof.stop ();
-    if lock_tracer s then begin
-      try Trace.Writer.close s.trace with e -> mark_failed s e
-    end;
-    curr_active_tracer := None
+    Mutex.protect s.mutex (fun () ->
+        try Trace.Writer.close s.trace
+        with e ->
+          (s.failed <- true; s.report_exn e);
+        Atomic.set curr_active_tracer None
+      )
   end
 
 let[@inline never] ext_alloc_slowpath ~bytes =
-  match !curr_active_tracer with
-  | None -> bytes_before_ext_sample := max_int; None
+  match Atomic.get curr_active_tracer with
+  | None -> Atomic.set bytes_before_ext_sample max_int; None
   | Some s ->
-    if lock_tracer_ext s then begin
+    if lock_tracer s then begin
       match
         let bytes_per_word = Sys.word_size / 8 in
         (* round up to an integer number of words *)
         let size_words = (bytes + bytes_per_word - 1) / bytes_per_word in
-        let samples = ref 0 in
-        while !bytes_before_ext_sample <= 0 do
-          bytes_before_ext_sample :=
-            !bytes_before_ext_sample + draw_sampler_bytes s;
-          incr samples
+        let samples = Atomic.make 0 in
+        while Atomic.get bytes_before_ext_sample <= 0 do
+          Atomic.set bytes_before_ext_sample
+            (Atomic.get bytes_before_ext_sample + draw_sampler_bytes s);
```

Comment:
Here it seems safe to do `Atomic.get, set` but still `Atomic.fetch_and_add` is probably more efficient?

Reply:
Maybe since `bytes_before_ext_sample` can be mutated outside of the critical section (via `ext_alloc`) you actually should use `fetch_and_add` here too.